File: js/src/cst.mjs
// Universal lossless CST infrastructure for issue #138.
//
// This module implements the `.lino` concrete syntax tree (CST) model
// described in `docs/case-studies/issue-138/cst-model.md`. It is the shared
// core used by every host-language converter (Rust, JavaScript, Lean 4, Rocq):
// each converter produces a tree of `CstNode` values, which can be printed
// back to bytes via `printCst()`, giving a byte-faithful round-trip.
//
// Three node kinds, exactly as the model specifies:
//
// - `list`   — a list node tagged with a dialect-specific symbol (e.g.
//              `lino-cst.rust.fn`); children are emitted in order with the
//              `open` and `close` delimiters if present.
// - `token`  — a leaf carrying a single non-trivia lexeme; its `text` field is
//              the original source bytes.
// - `trivia` — a leaf carrying whitespace or a comment; `text` is the original
//              source bytes. Trivia attaches to the *following* non-trivia
//              leaf by convention (Roslyn/libsyntax style).
//
// The tree is intentionally minimal: each converter owns the choice of tags
// and the placement of trivia. The printer is content-agnostic — it simply
// concatenates `text` in document order.

/** @typedef {Object} CstListNode
 *  @property {'list'} kind
 *  @property {string|null} tag
 *  @property {string|null} open
 *  @property {string|null} close
 *  @property {CstNode[]} children
 */
/** @typedef {Object} CstTokenNode
 *  @property {'token'} kind
 *  @property {string|null} tag
 *  @property {string} text
 */
/** @typedef {Object} CstTriviaNode
 *  @property {'trivia'} kind
 *  @property {string|null} tag
 *  @property {string} text
 */
/** @typedef {CstListNode|CstTokenNode|CstTriviaNode} CstNode */
/** @typedef {Object} CstListOptions
 *  @property {string|null} [open]
 *  @property {string|null} [close]
 */

/**
 * Construct a `list` CST node.
 *
 * @param {string|null} tag dialect-specific symbol, e.g. `lino-cst.rust.fn`.
 * @param {CstNode[]} children child nodes in document order.
 * @param {CstListOptions} [opts]
 *   optional `open`/`close` delimiter strings to emit around the children.
 * @returns {CstListNode}
 */
export function list(tag, children = [], opts = {}) {
  return {
    kind: 'list',
    tag: tag === undefined ? null : tag,
    open: opts.open ?? null,
    close: opts.close ?? null,
    children: children.slice(),
  };
}

/**
 * Construct a `token` CST leaf.
 *
 * @param {string} text original source bytes for the lexeme.
 * @param {string|null} [tag] optional dialect-specific symbol.
 * @returns {CstTokenNode}
 */
export function token(text, tag = null) {
  return { kind: 'token', tag, text: String(text) };
}

/**
 * Construct a `trivia` CST leaf (whitespace or comment).
 *
 * @param {string} text original source bytes.
 * @param {string|null} [tag] optional categorisation, e.g. `comment.line`.
 * @returns {CstTriviaNode}
 */
export function trivia(text, tag = null) {
  return { kind: 'trivia', tag, text: String(text) };
}

/**
 * Print a CST node back to its original byte-for-byte source representation.
 *
 * The walk is the five-line algorithm from the model document:
 *   - for `token` and `trivia` nodes, emit `node.text`.
 *   - for `list` nodes, emit `open`, then children in order, then `close`.
 *
 * @param {CstNode} node tree to print.
 * @returns {string} reconstructed source.
 */
export function printCst(node) {
  const out = [];
  emit(node, out);
  return out.join('');
}

function emit(node, out) {
  if (!node) return;
  if (node.kind === 'token' || node.kind === 'trivia') {
    out.push(node.text);
    return;
  }
  if (node.kind === 'list') {
    if (node.open) out.push(node.open);
    for (const child of node.children) emit(child, out);
    if (node.close) out.push(node.close);
    return;
  }
  throw new TypeError(`Unknown CST node kind: ${node && node.kind}`);
}

/**
 * Iterate every leaf (`token` or `trivia`) of a CST in document order.
 *
 * @param {CstNode} node tree to walk.
 * @returns {Iterable<CstTokenNode|CstTriviaNode>}
 */
export function* leaves(node) {
  if (!node) return;
  if (node.kind === 'token' || node.kind === 'trivia') {
    yield node;
    return;
  }
  if (node.kind === 'list') {
    for (const child of node.children) yield* leaves(child);
  }
}

/**
 * Serialise a CST node into a `.lino` S-expression suitable for embedding in
 * `.lino` files or for round-trip testing. The serialisation is the literal
 * shape documented in `docs/case-studies/issue-138/cst-model.md`:
 *
 *   `(lino-cst.list <tag> <child> <child> ...)`
 *   `(lino-cst.token "<text>")`
 *   `(lino-cst.trivia "<text>")`
 *
 * String contents are escaped using JSON rules so that the result is a valid
 * LiNo expression. `parseCstLino` is the exact inverse.
 *
 * @param {CstNode} node tree to serialise.
 * @returns {string}
 */
export function cstToLino(node) {
  if (!node) return '';
  if (node.kind === 'token') return `(lino-cst.token ${escapeText(node.text)})`;
  if (node.kind === 'trivia') return `(lino-cst.trivia ${escapeText(node.text)})`;
  if (node.kind === 'list') {
    const parts = ['lino-cst.list'];
    if (node.tag) parts.push(node.tag);
    if (node.open !== null && node.open !== undefined) parts.push(`(open ${escapeText(node.open)})`);
    if (node.close !== null && node.close !== undefined) parts.push(`(close ${escapeText(node.close)})`);
    for (const child of node.children) parts.push(cstToLino(child));
    return `(${parts.join(' ')})`;
  }
  throw new TypeError(`Unknown CST node kind: ${node && node.kind}`);
}

function escapeText(text) {
  return JSON.stringify(String(text));
}

function unescapeText(literal) {
  return JSON.parse(literal);
}

/**
 * Parse the `.lino` S-expression produced by `cstToLino()` back into a
 * `CstNode` tree. Inverse of `cstToLino`.
 *
 * @param {string} src `.lino` S-expression.
 * @returns {CstNode}
 */
export function linoToCst(src) {
  const tokens = tokeniseLinoCst(String(src));
  let i = 0;
  function peek() { return tokens[i]; }
  function eat() { return tokens[i++]; }
  function expect(t) {
    const got = eat();
    if (got !== t) throw new SyntaxError(`expected ${JSON.stringify(t)}, got ${JSON.stringify(got)}`);
  }
  function parseNode() {
    expect('(');
    const head = eat();
    if (head === 'lino-cst.token') {
      const lit = eat();
      expect(')');
      return token(unescapeText(lit));
    }
    if (head === 'lino-cst.trivia') {
      const lit = eat();
      expect(')');
      return trivia(unescapeText(lit));
    }
    if (head === 'lino-cst.list') {
      let tag = null;
      let open = null;
      let close = null;
      const children = [];
      while (peek() !== ')') {
        if (peek() === '(') {
          const lookahead = tokens[i + 1];
          if (lookahead === 'open') {
            eat(); eat();
            open = unescapeText(eat());
            expect(')');
            continue;
          }
          if (lookahead === 'close') {
            eat(); eat();
            close = unescapeText(eat());
            expect(')');
            continue;
          }
          children.push(parseNode());
        } else {
          if (tag !== null) throw new SyntaxError(`unexpected token ${peek()}`);
          tag = eat();
        }
      }
      expect(')');
      return list(tag, children, { open, close });
    }
    throw new SyntaxError(`unknown CST tag: ${head}`);
  }
  return parseNode();
}

function tokeniseLinoCst(src) {
  const out = [];
  let i = 0;
  while (i < src.length) {
    const c = src[i];
    if (c === '(' || c === ')') {
      out.push(c);
      i++;
      continue;
    }
    if (c === ' ' || c === '\t' || c === '\n' || c === '\r') {
      i++;
      continue;
    }
    if (c === '"') {
      let j = i + 1;
      while (j < src.length) {
        if (src[j] === '\\') { j += 2; continue; }
        if (src[j] === '"') break;
        j++;
      }
      if (j >= src.length) throw new SyntaxError('unterminated string literal');
      out.push(src.substring(i, j + 1));
      i = j + 1;
      continue;
    }
    let j = i;
    while (j < src.length && !' \t\n\r()'.includes(src[j])) j++;
    out.push(src.substring(i, j));
    i = j;
  }
  return out;
}

/**
 * The four host-language dialect tag prefixes plus the shared dialect.
 * Exported so that external tooling can refer to them symbolically.
 */
export const DIALECTS = Object.freeze({
  rust: 'lino-cst.rust',
  js: 'lino-cst.js',
  lean: 'lino-cst.lean',
  rocq: 'lino-cst.rocq',
  shared: 'lino-cst.shared',
});

/**
 * Compute and return a structural copy of a CST node. Used by translators that
 * need to mutate without altering the input.
 *
 * @param {CstNode} node
 * @returns {CstNode}
 */
export function cloneCst(node) {
  if (!node) return node;
  if (node.kind === 'token' || node.kind === 'trivia') {
    return { ...node };
  }
  return {
    kind: 'list',
    tag: node.tag,
    open: node.open,
    close: node.close,
    children: node.children.map(cloneCst),
  };
}

File: rust/src/cst.rs
//! Universal lossless CST infrastructure for issue #138.
//!
//! This module mirrors `js/src/cst.mjs`: three node kinds (`list`, `token`,
//! `trivia`), a content-agnostic round-trip printer, and `.lino`
//! serialisation/deserialisation helpers. Used by `cst_rust`, `cst_js`,
//! `cst_lean` and `cst_rocq` to express their token streams.

use std::fmt::Write;

/// The four host-language dialect tag prefixes plus the shared dialect.
pub mod dialects {
    pub const RUST: &str = "lino-cst.rust";
    pub const JS: &str = "lino-cst.js";
    pub const LEAN: &str = "lino-cst.lean";
    pub const ROCQ: &str = "lino-cst.rocq";
    pub const SHARED: &str = "lino-cst.shared";
}

/// A CST node. Three kinds: `List` (with optional `open`/`close` delimiters),
/// `Token` (significant lexeme), `Trivia` (whitespace or comment).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CstNode {
    /// A list node with an optional dialect tag and optional open/close
    /// delimiter strings (e.g. `(`, `)`).
    List {
        tag: Option<String>,
        open: Option<String>,
        close: Option<String>,
        children: Vec<CstNode>,
    },
    /// A non-trivia lexeme. `text` holds the original source bytes.
    Token { tag: Option<String>, text: String },
    /// Whitespace or comment trivia. `text` holds the original source bytes.
    Trivia { tag: Option<String>, text: String },
}

impl CstNode {
    /// Construct a `List` node.
    pub fn list(tag: impl Into<String>, children: Vec<CstNode>) -> Self {
        CstNode::List {
            tag: Some(tag.into()),
            open: None,
            close: None,
            children,
        }
    }

    /// Construct an untagged `List` node with custom open/close delimiters.
    pub fn list_with_delims(
        tag: Option<String>,
        open: Option<String>,
        close: Option<String>,
        children: Vec<CstNode>,
    ) -> Self {
        CstNode::List {
            tag,
            open,
            close,
            children,
        }
    }

    /// Construct a `Token` leaf.
    pub fn token(text: impl Into<String>, tag: Option<&str>) -> Self {
        CstNode::Token {
            tag: tag.map(String::from),
            text: text.into(),
        }
    }

    /// Construct a `Trivia` leaf.
    pub fn trivia(text: impl Into<String>, tag: Option<&str>) -> Self {
        CstNode::Trivia {
            tag: tag.map(String::from),
            text: text.into(),
        }
    }

    /// True if this node is a list.
    pub fn is_list(&self) -> bool {
        matches!(self, CstNode::List { .. })
    }

    /// The dialect tag, if any.
    pub fn tag(&self) -> Option<&str> {
        match self {
            CstNode::List { tag, .. } => tag.as_deref(),
            CstNode::Token { tag, .. } => tag.as_deref(),
            CstNode::Trivia { tag, .. } => tag.as_deref(),
        }
    }

    /// The textual content of a leaf node. Returns `""` for lists.
    pub fn text(&self) -> &str {
        match self {
            CstNode::Token { text, .. } => text,
            CstNode::Trivia { text, .. } => text,
            CstNode::List { .. } => "",
        }
    }

    /// Iterate every leaf (token/trivia) of this subtree in document order.
    pub fn leaves(&self) -> Vec<&CstNode> {
        let mut out = Vec::new();
        collect_leaves(self, &mut out);
        out
    }
}

fn collect_leaves<'a>(node: &'a CstNode, out: &mut Vec<&'a CstNode>) {
    match node {
        CstNode::List { children, .. } => {
            for c in children {
                collect_leaves(c, out);
            }
        }
        _ => out.push(node),
    }
}

/// Print a CST node back to its byte-faithful source form.
pub fn print_cst(node: &CstNode) -> String {
    let mut out = String::new();
    emit(node, &mut out);
    out
}

fn emit(node: &CstNode, out: &mut String) {
    match node {
        CstNode::Token { text, .. } | CstNode::Trivia { text, .. } => {
            out.push_str(text);
        }
        CstNode::List {
            children,
            open,
            close,
            ..
        } => {
            if let Some(o) = open {
                out.push_str(o);
            }
            for c in children {
                emit(c, out);
            }
            if let Some(c) = close {
                out.push_str(c);
            }
        }
    }
}

/// Serialise a CST node into a `.lino` S-expression matching the format
/// produced by `js/src/cst.mjs`'s `cstToLino`.
pub fn cst_to_lino(node: &CstNode) -> String {
    let mut out = String::new();
    write_lino(node, &mut out);
    out
}

fn write_lino(node: &CstNode, out: &mut String) {
    match node {
        CstNode::Token { text, .. } => {
            let _ = write!(out, "(lino-cst.token {})", escape_text(text));
        }
        CstNode::Trivia { text, .. } => {
            let _ = write!(out, "(lino-cst.trivia {})", escape_text(text));
        }
        CstNode::List {
            tag,
            open,
            close,
            children,
        } => {
            out.push('(');
            out.push_str("lino-cst.list");
            if let Some(t) = tag {
                out.push(' ');
                out.push_str(t);
            }
            if let Some(o) = open {
                let _ = write!(out, " (open {})", escape_text(o));
            }
            if let Some(c) = close {
                let _ = write!(out, " (close {})", escape_text(c));
            }
            for c in children {
                out.push(' ');
                write_lino(c, out);
            }
            out.push(')');
        }
    }
}

fn escape_text(text: &str) -> String {
    let mut out = String::with_capacity(text.len() + 2);
    out.push('"');
    for ch in text.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\x08' => out.push_str("\\b"),
            '\x0C' => out.push_str("\\f"),
            c if (c as u32) < 0x20 => {
                let _ = write!(out, "\\u{:04x}", c as u32);
            }
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

/// Parse the `.lino` S-expression produced by `cst_to_lino` back into a
/// `CstNode`. Inverse of `cst_to_lino`.
pub fn lino_to_cst(src: &str) -> Result<CstNode, String> {
    let tokens = tokenise_lino_cst(src)?;
    let mut idx = 0usize;
    let node = parse_node(&tokens, &mut idx)?;
    Ok(node)
}

fn parse_node(tokens: &[String], idx: &mut usize) -> Result<CstNode, String> {
    expect(tokens, idx, "(")?;
    let head = eat(tokens, idx)?;
    match head.as_str() {
        "lino-cst.token" => {
            let lit = eat(tokens, idx)?;
            expect(tokens, idx, ")")?;
            Ok(CstNode::token(unescape_text(&lit)?, None))
        }
        "lino-cst.trivia" => {
            let lit = eat(tokens, idx)?;
            expect(tokens, idx, ")")?;
            Ok(CstNode::trivia(unescape_text(&lit)?, None))
        }
        "lino-cst.list" => {
            let mut tag: Option<String> = None;
            let mut open: Option<String> = None;
            let mut close: Option<String> = None;
            let mut children: Vec<CstNode> = Vec::new();
            while peek(tokens, *idx)? != ")" {
                if peek(tokens, *idx)? == "(" {
                    let lookahead = peek(tokens, *idx + 1)?;
                    if lookahead == "open" {
                        *idx += 2;
                        let lit = eat(tokens, idx)?;
                        expect(tokens, idx, ")")?;
                        open = Some(unescape_text(&lit)?);
                        continue;
                    }
                    if lookahead == "close" {
                        *idx += 2;
                        let lit = eat(tokens, idx)?;
                        expect(tokens, idx, ")")?;
                        close = Some(unescape_text(&lit)?);
                        continue;
                    }
                    children.push(parse_node(tokens, idx)?);
                } else {
                    if tag.is_some() {
                        return Err(format!("unexpected token {:?}", peek(tokens, *idx)?));
                    }
                    tag = Some(eat(tokens, idx)?);
                }
            }
            expect(tokens, idx, ")")?;
            Ok(CstNode::list_with_delims(tag, open, close, children))
        }
        other => Err(format!("unknown CST tag: {}", other)),
    }
}

fn peek(tokens: &[String], idx: usize) -> Result<&str, String> {
    tokens
        .get(idx)
        .map(String::as_str)
        .ok_or_else(|| "unexpected end of input".to_string())
}

fn eat(tokens: &[String], idx: &mut usize) -> Result<String, String> {
    let s = tokens
        .get(*idx)
        .cloned()
        .ok_or_else(|| "unexpected end of input".to_string())?;
    *idx += 1;
    Ok(s)
}

fn expect(tokens: &[String], idx: &mut usize, t: &str) -> Result<(), String> {
    let got = eat(tokens, idx)?;
    if got != t {
        return Err(format!("expected {:?}, got {:?}", t, got));
    }
    Ok(())
}

fn tokenise_lino_cst(src: &str) -> Result<Vec<String>, String> {
    let bytes: Vec<char> = src.chars().collect();
    let mut out = Vec::new();
    let mut i = 0usize;
    while i < bytes.len() {
        let c = bytes[i];
        if c == '(' || c == ')' {
            out.push(c.to_string());
            i += 1;
            continue;
        }
        if c == ' ' || c == '\t' || c == '\n' || c == '\r' {
            i += 1;
            continue;
        }
        if c == '"' {
            let mut j = i + 1;
            while j < bytes.len() {
                if bytes[j] == '\\' {
                    j += 2;
                    continue;
                }
                if bytes[j] == '"' {
                    break;
                }
                j += 1;
            }
            if j >= bytes.len() {
                return Err("unterminated string literal".to_string());
            }
            out.push(bytes[i..=j].iter().collect());
            i = j + 1;
            continue;
        }
        let mut j = i;
        while j < bytes.len() && !" \t\n\r()".contains(bytes[j]) {
            j += 1;
        }
        out.push(bytes[i..j].iter().collect());
        i = j;
    }
    Ok(out)
}

fn unescape_text(literal: &str) -> Result<String, String> {
    let chars: Vec<char> = literal.chars().collect();
    if chars.first() != Some(&'"') || chars.last() != Some(&'"') {
        return Err(format!("not a string literal: {}", literal));
    }
    let mut out = String::new();
    let mut i = 1;
    while i < chars.len() - 1 {
        let c = chars[i];
        if c == '\\' {
            i += 1;
            if i >= chars.len() - 1 {
                return Err("trailing backslash".to_string());
            }
            let esc = chars[i];
            match esc {
                '"' => out.push('"'),
                '\\' => out.push('\\'),
                '/' => out.push('/'),
                'n' => out.push('\n'),
                'r' => out.push('\r'),
                't' => out.push('\t'),
                'b' => out.push('\x08'),
                'f' => out.push('\x0C'),
                'u' => {
                    if i + 4 >= chars.len() {
                        return Err("bad \\u escape".to_string());
                    }
                    let hex: String = chars[i + 1..=i + 4].iter().collect();
                    let cp = u32::from_str_radix(&hex, 16)
                        .map_err(|e| format!("bad \\u escape: {}", e))?;
                    if let Some(ch) = char::from_u32(cp) {
                        out.push(ch);
                    }
                    i += 4;
                }
                other => return Err(format!("unknown escape: \\{}", other)),
            }
            i += 1;
        } else {
            out.push(c);
            i += 1;
        }
    }
    Ok(out)
}

/// Convenience: return a clone of `node`. Provided for parity with the JS
/// module which exports `cloneCst` explicitly.
pub fn clone_cst(node: &CstNode) -> CstNode {
    node.clone()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prints_list_with_open_close() {
        let n = CstNode::list_with_delims(
            Some("demo".into()),
            Some("(".into()),
            Some(")".into()),
            vec![CstNode::token("a", None), CstNode::token("b", None)],
        );
        assert_eq!(print_cst(&n), "(ab)");
    }

    #[test]
    fn lino_round_trip_for_token_trivia_list() {
        let n = CstNode::list(
            "lino-cst.rust.fn",
            vec![
                CstNode::token("fn ", Some("kw")),
                CstNode::token("foo", Some("ident")),
                CstNode::trivia(" ", Some("ws")),
            ],
        );
        let s = cst_to_lino(&n);
        let back = lino_to_cst(&s).unwrap();
        assert_eq!(print_cst(&n), print_cst(&back));
    }
}
